### ISSUE TITLE
ccache fails when switching from compiler name to a full path

### ISSUE THREAD

**Issue by hebasto** [ccache/ccache#1326]
ccache works fine when a compiler is specified by its name:
```
$ ccache -z && ccache env -u VAR clang -c -o main.o main.cpp && ccache -sv
Statistics zeroed
Cache directory:    /home/hebasto/.cache/ccache
Config file:        /home/hebasto/.config/ccache/ccache.conf
System config file: /usr/local/etc/ccache.conf
Stats updated:      Fri Aug 25 12:08:19 2023
Cacheable calls:        1 /   1 (100.0%)
  Hits:                 0 /   1 ( 0.00%)
    Direct:             0
    Preprocessed:       0
  Misses:               1 /   1 (100.0%)
Successful lookups:
  Direct:               0 /   1 ( 0.00%)
  Preprocessed:         0 /   1 ( 0.00%)
Local storage:
  Cache size (GiB):   4.2 / 5.0 (83.59%)
  Files:            18564
  Hits:                 0 /   1 ( 0.00%)
  Misses:               1 /   1 (100.0%)
  Reads:                2
  Writes:               2
```

However, it fails with "Multiple source files" when the full path to the compiler is provided:
```
$ ccache -z && ccache env -u VAR /usr/bin/clang -c -o main.o main.cpp && ccache -sv
Statistics zeroed
Cache directory:         /home/hebasto/.cache/ccache
Config file:             /home/hebasto/.config/ccache/ccache.conf
System config file:      /usr/local/etc/ccache.conf
Stats updated:           Fri Aug 25 12:09:16 2023
Cacheable calls:             0 /   1 ( 0.00%)
  Hits:                      0
    Direct:                  0
    Preprocessed:            0
  Misses:                    0
Uncacheable calls:           1 /   1 (100.0%)
  Multiple source files:     1 /   1 (100.0%)
Successful lookups:
  Direct:                    0
  Preprocessed:              0
Local storage:
  Cache size (GiB):        4.2 / 5.0 (83.59%)
  Files:                 18564
  Hits:                      0
  Misses:                    0
  Reads:                     0
  Writes:                    0
```

**Comment by jrosdahl** [comment]
Ccache supports two forms for compilation as documented in [_Run modes_](https://ccache.dev/manual/4.8.2.html#_run_modes), where the `ccache compiler [compiler options]` form takes the compiler as the **first** argument to `ccache`. In your case that's `env`, which is not a supported compiler.

The reason why it _kinda-sorta_ works with `ccache env -u VAR clang ...` is that:

- `-u` is a [linker option](https://gcc.gnu.org/onlinedocs/gcc/Link-Options.html#index-u) understood by ccache: it consumes one argument, so `VAR` won't be considered as a potential source code file.
- If `clang` doesn't exist in the local directory it's also ruled out as a potential source code file.

ccache then succeeds in constructing a command line for the "compiler" `env`. Since `env` is not the name of a known compiler type, ccache guesses that it is GCC. However, this only sorta-kinda works since the [compiler check](https://ccache.dev/manual/4.8.2.html#config_compiler_check) by default uses the compiler's mtime to detect compiler upgrades. Including `env`'s mtime in the input hash will not detect clang version upgrades, so it's not a good idea to use that form. See also [_Using ccache with other compiler wrappers_](https://ccache.dev/manual/4.8.2.html#_using_ccache_with_other_compiler_wrappers).

When you use `ccache env -u VAR /usr/bin/clang ...` then ccache sees that `/usr/bin/clang` exists and therefore concludes that it is a source code file to pass to the compiler `env`. Then `main.cpp` is found as the second source code file, and ccache doesn't support passing multiple source code files to the compiler.

In summary: Use the form `env -u VAR ccache clang ...` instead.
